### ISSUE TITLE
Invalid used-before-assignment

### ISSUE THREAD

**Issue by socketpair** [PyCQA/pylint#5112]
### Bug description

```python
from random import randint

asd2 = {
    **({'param': x} if (x := randint(0, 1)) else {}),
}
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
E0601: Using variable 'x' before assignment (used-before-assignment)
```

### Expected behavior

no error.

### Pylint version

```shell
pylint 2.9.6
astroid 2.6.6
Python 3.9.7 (default, Aug 30 2021, 00:00:00) 
[GCC 11.2.1 20210728 (Red Hat 11.2.1-1)]
```

### OS / Environment

Fedora 34

**Comment by socketpair** [comment]
```python
from random import randint

def asd():
    return [dom] if (dom := randint(-1, 2)) else []
```

The same

**Comment by natgiordano** [comment]
Hi, I'd like to take a look at this issue.

**Comment by Pierre-Sassoulas** [comment]
Neat, I assigned you to the issue, to not hesitate to ask for help here or in a draft merge request :)

**Comment by natgiordano** [comment]
Hey, could you actually unassign me from this issue? I meant to request #5187 and accidentally requested this. I went ahead and commented on #5187 like I initially intended to. Sorry for the confusion!
